cpp/src/buffer/resource.cpp

```diff
@@ -136,9 +123,12 @@ std::unique_ptr<rmm::device_buffer> BufferResource::move_to_device_buffer(
     rmm::cuda_stream_view stream,
     MemoryReservation& reservation
 ) {
-    return std::move(
-        move(MemoryType::DEVICE, std::move(buffer), stream, reservation)->device()
+    RAPIDSMPF_EXPECTS(
+        reservation.mem_type_ == MemoryType::DEVICE,
+        "the memory type of MemoryReservation doesn't match",
+        std::invalid_argument
     );
+    return std::move(move(std::move(buffer), stream, reservation)->device());
```

Comment:
```suggestion
    return move(std::move(buffer), stream, reservation)->device();
```

Reply:
This move is required because  `..->device()` is a ref, but I agree its clunky. I added a separate `release_device()` method to make it bit more "less clunky" :D